src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs

```diff
@@ -40,15 +40,71 @@ public override Stream CreateStream(DiagnosticBag diagnostics)
 
                 try
                 {
-                    return _streamToDispose = _compiler.FileOpen(_filePath, PortableShim.FileMode.Create, PortableShim.FileAccess.ReadWrite, PortableShim.FileShare.None);
+                    try
+                    {
+                        return OpenFileStream();
+                    }
+                    catch (IOException e)
+                    {
+                        // Other process is reading the file preventing us to write to it.
+                        // We attempt to rename and delete the file in case the reader opened it with FileShare.Delete flag that
+                        // allows the file to be deleted by other processes.
+                        //
+                        // Note that if the file is marked "readonly" or the current user doesn't have sufficient privileges
+                        // the exception thrown is UnauthorizedAccessException, not IOException, so we won't attempt to delete the file.
+
+                        try
+                        {
+                            const int eWin32SharingViolation = unchecked((int)0x80070020);
+
+                            if (PathUtilities.IsUnixLikePlatform)
+                            {
+                                // Unix & Mac are simple: just delete the file in the directory. 
+                                // The memory mapped content remains available for the reader.
+                                PortableShim.File.Delete(_filePath);
```

Comment:
Last comment (I swear :smile:).  Does this change only apply to memory mapped scenarios on Unix?  Unix, to my knowledge, doesn't have the `FileShare.Delete` option hence couldn't hit the primary case listed by this bug.  

CC @stephentoub